### PR TITLE
ADR for Securing JSON Output

### PR DESCRIPTION
**Pull request by tommotaylor**
ADR for securing the JSON output payload.

STATUS: Proposed

See formatted markdown here: https://github.com/ministryofjustice/form-builder/blob/json-output-security-adr/decisions/0002-secure-JSON-output.md

Jira card, containing some discussion: https://dsdmoj.atlassian.net/jira/software/projects/FB/boards/323?selectedIssue=FB-525

### CODE REVIEW

File: decisions/0002-secure-JSON-output.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+- Signs JSON payload using a shared secret and JWS protocol
+- Sends as POST request
+
+|
+V
+
+Adapter:
+- Receives HTTPS POST request from Form Builder
+- Validates JWS signature using shared secret
+- Also verifies request header / origin to ensure it was sent from Form Builder (edited
```
Comment (asmega) [inline decisions/0002-secure-JSON-output.md]: minor typo. minus `(edited`

**Comment by tommotaylor** [comment]
@jennyd Has suggested the following changes after discussion with Joel Samuel:
- One way TLS over mutual TLS
- Payload encryption over signature

Discussion from Jenny:
```
We came to the conclusion that mutual TLS doesn’t really bring any benefits and involves way too much management overhead in this context, so agreed on normal TLS + payload encryption with a pre-shared secret
The reason for the payload encryption is that TLS is often terminated at the edge of a large network and then travels in the clear inside that network, and given the sensitivity of the submitted data for some forms and that it may need to be relied on in court, we need more confidence that only the intended recipient can read the data and that other actors on that network can’t impersonate FB by sending other requests to the endpoint.
(We debated whether signing the payload is also necessary for the last point about impersonation, but decided that encryption with a pre-shared secret basically gives us enough protection against that - encryption for a public key that’s treated as public wouldn’t though)
```

File: decisions/0002-secure-JSON-output.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+Use TLS over HTTP (HTTPS). This would ensure
+encrypted data in transit, which [satisfies MOJ standards](https://ministryofjustice.github.io/security-guidance/standards/cryptography/#cryptography).
+We came to the conclusion that mutual TLS doesn’t really bring any benefits and
+involves way too much management overhead in this context, so agreed on normal
+TLS + payload encryption with a pre-shared secret.
+
+### Encrypted JSON Payload
+Encryption of the payload is possible with JWE (JOSE toolkit). This would require
+decryption at the other end. Payload encryption is preferred over TLS alone as TLS
+is often terminated at the edge of a large network with ommunications travelling in the clear
```
Comment (jennyd) [inline decisions/0002-secure-JSON-output.md]: typo: `ommunications`

File: decisions/0002-secure-JSON-output.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+inside that network. Given the sensitivity of the submitted data for some forms
+and that it may need to be relied on in court, we need more confidence that only
+the intended recipient can read the data and that other actors on that network
+can’t impersonate FB by sending other requests to the endpoint.
+
+### Shared Secret
+The shared secret will be set as an ENV var in the publisher to be consumed by
```
Comment (jennyd) [inline decisions/0002-secure-JSON-output.md]: Using a normal envvar is ok right now as we're building the receiving app for the first use case, but we should move to the Publisher generating the secret for the form owner to copy out into their config. That will give us confidence that the secrets are appropriate for the encryption method we choose.

File: decisions/0002-secure-JSON-output.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+Form Builder:
+- Connects with adapter via HTTPS using a ruby library such as [Net:HTTP](https://ruby-doc.org/stdlib-2.6.3/libdoc/net/http/rdoc/Net/HTTP.html).
+- Connection with TLS using `Net::HTTP` gem or similar.
+- Encrypts JSON payload using a shared secret and JWE protocol.
+- Sends as POST request.
+
+Adapter:
+- Receives HTTPS POST request from Form Builder.
+- Decrypts JSON payload using shared secret.
+- Also verifies request header / origin to ensure it was sent from Form Builder.
```
Comment (jennyd) [inline decisions/0002-secure-JSON-output.md]: The payload being encrypted with the shared secret should be sufficient for knowing the request comes from FB, I'm not sure there's any value in doing anything else.

File: decisions/0002-secure-JSON-output.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+generated to ensure that it is appropriate for the encryption method we choose.
+
+### Certificates vs Shared Secret
+Both certificates and shared secrets can be used for signing and de-serialising the payload.
+Given users may not be technical, a shared secret would be preferred as requiring users
+to generate and upload certificates may be too much to ask.
+
+### Validation of remote endpoint
+The Form Builder system should validate the endpoint that is entered by the user.
+At minimum this should an an HTTPS endpoint and `*.gov.uk`.
```
Comment (jennyd) [inline decisions/0002-secure-JSON-output.md]: typo: `an an`

**Pull request merged**